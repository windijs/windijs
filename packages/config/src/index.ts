export * from "./bg";
export * from "./svg";
export * from "./flex";
export * from "./grid";
export * from "./border";
export * from "./effects";
export * from "./behaviors";
export * from "./filter";
export * from "./opacity";
export * from "./sizing";
export * from "./spacing";
export * from "./animation";
export * from "./transform";
export * from "./positioning";
export * from "./interactivity";
export * from "./transitions";
export * from "./typography";
export * from "./keyframes";
export * from "./interface";
