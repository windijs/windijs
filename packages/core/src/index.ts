export * from "./base";
export * from "./api";
export * from "./handler";
export * from "./prefixer";
export * from "./builder";
