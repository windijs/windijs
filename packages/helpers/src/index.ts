// TODO: maybe add more helpers reference sass?
// https://sass-lang.com/documentation/modules

export * from "./types";
export * from "./build";
export * from "./config";
export * from "./color";
export * from "./common";
export * from "./css";
export * from "./data";
export * from "./dollar";
export * from "./escape";
export * from "./funcs";
export * from "./math";
export * from "./meta";
export * from "./namer";
export * from "./proxy";
export * from "./unit";
