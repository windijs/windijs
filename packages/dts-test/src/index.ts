import { defineConfig } from "@windijs/config";

defineConfig({
  utilities: {

  },
});
