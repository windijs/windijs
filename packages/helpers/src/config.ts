import { Handler, StyleLoader, StyleObject } from "./types";

import { Theme } from "@windijs/config";

export type DarkModeConfig = "class" | "media" | false;

export type UtilitiesConfig = { [key: string]: Handler<unknown> | Array<Handler<unknown>> | StyleObject | UtilitiesConfig };

export type VariantsConfig = { [key: string]: string | string[] };

export interface Config<T extends object = Theme> {
  // presets?: Config[];
  // prefixer?: boolean;
  // attributify?: boolean | {
  //   prefix?: string
  //   separator?: string
  //   disable?: string[]
  // }
  separator?: string;
  important?: boolean | string;
  darkMode?: DarkModeConfig;
  theme?: T;
  // variantOrder?: string[];
  // plugins?: Plugin[];
  // handlers?: Handlers;
  // corePlugins?: (keyof BaseTheme)[] | string[] | { [ T in keyof BaseTheme ] : boolean } | { [ key:string ] : boolean };
  // prefix?: string;
  // exclude?: RegExp[];
  // alias?: { [key:string]: string };
  // shortcuts?: { [key:string]: Shortcut };
  // purge?: unknown;
  utilities?: UtilitiesConfig;
  variants?: VariantsConfig;
  styleLoader?: StyleLoader;
  [key:string]: any;
}

export function defineConfig (config: Config) {
  return config;
}
