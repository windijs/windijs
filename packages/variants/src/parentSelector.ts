import { createVariant } from "@windi/core";

/**
 * Targets an element when a marked parent matches the hover pseudo-class.
 */
export const groupHover = createVariant(".group:hover &");

/**
 * Targets an element when a marked parent matches the focus pseudo-class.
 */
export const groupFocus = createVariant(".group:focus &");

/**
 * Targets an element when a marked parent matches the active pseudo-class.
 */
export const groupActive = createVariant(".group:active &");

/**
 * Targets an element when a marked parent matches the visited pseudo-class.
 */
export const groupVisited = createVariant(".group:visited &");
