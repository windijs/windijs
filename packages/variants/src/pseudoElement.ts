import { createVariant } from "@windi/core";

/**
 * Represents a styleable child pseudo-element immediately after the originating element’s actual content.
 *
 * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/::after)
 */
export const after = createVariant("&::after");

/**
 * Used to create a backdrop that hides the underlying document for an element in a top layer (such as an element that is displayed fullscreen).
 *
 * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/::backdrop)
 */
// export const backdrop = createVariant("&::backdrop");

/**
 * Represents a styleable child pseudo-element immediately before the originating element’s actual content.
 *
 * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/::before)
 */
export const before = createVariant("&::before");

/**
 * Represents the first letter of an element, if it is not preceded by any other content (such as images or inline tables) on its line.
 *
 * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/::first-letter)
 */
export const firstLetter = createVariant("&::first-letter");

/**
 * Describes the contents of the first formatted line of its originating element.
 *
 * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/::first-line)
 */
export const firstLine = createVariant("&::first-line");

/**
 * The ::marker CSS pseudo-element selects the marker box of a list item, which typically contains a bullet or number. It works on any element or pseudo-element set to display: list-item, such as the \<li> and \<summary> elements.
 *
 * (Edge 86, Firefox 68, Safari 11.1, Chrome 86, Opera 72)
 *
 * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/::marker)
 */
export const marker = createVariant("& *::marker, &::marker");

/**
 * Represents the portion of a document that has been highlighted by the user.
 *
 * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/::selection)
 */
export const selection = createVariant("& *::selection, &::selection");
