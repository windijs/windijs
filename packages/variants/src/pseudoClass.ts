import { createVariant } from "@windijs/core";
import { useProxy } from "@windijs/helpers";
import { camelToDash } from "@windijs/shared";

import type { VariantBuilder } from "@windijs/helpers";

// TODO: the document of variants need to be improved.

interface PseudoClassVariants {
  /**
   * Represents user interface elements that are in a disabled state; such elements have a corresponding enabled state.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:disabled)
   */
  disabled: VariantBuilder;
  /**
   * Applies once the link has been visited by the user.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:visited)
   */
  visited: VariantBuilder;
  /**
   * Radio and checkbox elements can be toggled by the user. Some menu items are 'checked' when the user selects them. When such elements are toggled 'on' the :checked pseudo-class applies.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:checked)
   */
  checked: VariantBuilder;
  /**
   * The :focus-within pseudo-class applies to any element for which the :focus pseudo class applies as well as to an element whose descendant in the flat tree (including non-element nodes, such as text nodes) matches the conditions for matching :focus.
   *
   * (Edge 79, Firefox 52, Safari 10.1, Chrome 60, Opera 47)
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:focus-within)
   */
  focusWithin: VariantBuilder;
  /**
   * Applies while the user designates an element with a pointing device, but does not necessarily activate it. For example, a visual user agent could apply this pseudo-class when the cursor (mouse pointer) hovers over a box generated by the element.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:hover)
   */
  hover: VariantBuilder;
  /**
   * Applies while an element has the focus (accepts keyboard or mouse events, or other forms of input).
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:focus)
   */
  focus: VariantBuilder;
  /**
   * The :focus-visible pseudo-class applies while an element matches the :focus pseudo-class and the UA determines via heuristics that the focus should be made evident on the element.
   *
   * (Edge 86, Firefox 85, Safari 15.4, Chrome 86, Opera 72)
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:focus-visible)
   */
  focusVisible: VariantBuilder;
  /**
   * Applies while an element is being activated by the user. For example, between the times the user presses the mouse button and releases it.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:active)
   */
  active: VariantBuilder;
  /**
   * Applies to links that have not yet been visited.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:link)
   */
  link: VariantBuilder;
  /**
   * Some URIs refer to a location within a resource. This kind of URI ends with a 'number sign' (#) followed by an anchor identifier (called the fragment identifier).
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:target)
   */
  target: VariantBuilder;
  /**
   * Same as :not(:checked)
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:checked)
   */
  notChecked: VariantBuilder;
  /**
   * Represents user interface elements that are in an enabled state; such elements have a corresponding disabled state.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:enabled)
   */
  enabled: VariantBuilder;
  /**
   * Applies to UI elements whose value is in an indeterminate state.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:indeterminate)
   */
  indeterminate: VariantBuilder;
  /**
   * An element is :valid or :invalid when it is, respectively, valid or invalid with respect to data validity semantics defined by a different specification.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:invalid)
   */
  invalid: VariantBuilder;
  /**
   * An element is :valid or :invalid when it is, respectively, valid or invalid with respect to data validity semantics defined by a different specification.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:valid)
   */
  valid: VariantBuilder;
  /**
   * A form element is :required or :optional if a value for it is, respectively, required or optional before the form it belongs to is submitted. Elements that are not form elements are neither required nor optional.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:optional)
   */
  optional: VariantBuilder;
  /**
   * A form element is :required or :optional if a value for it is, respectively, required or optional before the form it belongs to is submitted. Elements that are not form elements are neither required nor optional.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:required)
   */
  required: VariantBuilder;
  /**
   * ⚠️ Property is experimental. Be cautious when using it.️
   *
   * The :placeholder-shown CSS pseudo-class represents any \<input> or \<textarea> element that is currently displaying placeholder text.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:placeholder-shown)
   */
  placeholderShown: VariantBuilder;
  /**
   * An element whose contents are not user-alterable is :read-only. However, elements whose contents are user-alterable (such as text input fields) are considered to be in a :read-write state. In typical documents, most elements are :read-only.
   *
   * (Edge 13, Firefox 78, Safari 4, Chrome 1, Opera 9)
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:read-only)
   */
  readOnly: VariantBuilder;
  /**
   * An element whose contents are not user-alterable is :read-only. However, elements whose contents are user-alterable (such as text input fields) are considered to be in a :read-write state. In typical documents, most elements are :read-only.
   *
   * (Edge 13, Firefox 78, Safari 4, Chrome 1, Opera 9)
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:read-write)
   */
  readWrite: VariantBuilder;
  /**
   * Same as :not(:disabled)
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:disabled)
   */
  notDisabled: VariantBuilder;
  /**
   * Same as :nth-of-type(1). Represents an element that is the first sibling of its type in the list of children of its parent element.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:first-of-type)
   */
  firstOfType: VariantBuilder;
  /**
   * Same as :not(:nth-of-type(1))
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:first-of-type)
   */
  notFirstOfType: VariantBuilder;
  /**
   * Same as :nth-last-of-type(1). Represents an element that is the last sibling of its type in the list of children of its parent element.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:last-of-type)
   */
  lastOfType: VariantBuilder;
  /**
   * Same as :not(:last-of-type)
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:last-of-type)
   */
  notLastOfType: VariantBuilder;
  /**
   * Represents an element that has a parent element and whose parent element has no other element children. Same as :first-child:last-child or :nth-child(1):nth-last-child(1), but with a lower specificity.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:only-child)
   */
  onlyChild: VariantBuilder;
  /**
   * Same as :not(:only-child)
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:only-child)
   */
  notOnlyChild: VariantBuilder;
  /**
   * Matches every element that is the only child of its type, of its parent. Same as :first-of-type:last-of-type or :nth-of-type(1):nth-last-of-type(1), but with a lower specificity.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:only-of-type)
   */
  onlyOfType: VariantBuilder;
  /**
   * Same as :not(:only-of-type)
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:only-of-type)
   */
  notOnlyOfType: VariantBuilder;
  /**
   * Represents an element that is the root of the document. In HTML 4, this is always the HTML element.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:root)
   */
  root: VariantBuilder;
  /**
   * Represents an element that has no children at all.
   *
   * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:empty)
   */
  empty: VariantBuilder;
}

/**
 * Same as :nth-child(1). Represents an element that is the first child of some other element.
 *
 * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:first-child)
 */
export const first = createVariant("&:first-child");

/**
 * Same as :nth-last-child(1). Represents an element that is the last child of some other element.
 *
 * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:last-child)
 */
export const last = createVariant("&:last-child");

/**
 * Same as :nth-of-type(even)
 *
 * [MDN Reference](https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-of-type)
 */
export const evenOfType = createVariant("&:nth-of-type(even)");

/**
 * Same as :nth-of-type(odd)
 *
 * [MDN Reference](https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-of-type)
 */
export const oddOfType = createVariant("&:nth-of-type(odd)");

/**
 * Applies to the one or more UI elements that are the default among a set of similar elements. Typically applies to context menu items, buttons, and select lists/menus.
 *
 * (Edge 79, Firefox 4, Safari 5, Chrome 10, Opera 10)
 *
 * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:default)
 */
export const $default = createVariant("&:default");

/**
 * Same as :not(:first-child).
 *
 * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:first-child)
 */
export const notFirst = createVariant("&:not(:first-child)");

/**
 * Same as :not(:last-child).
 *
 * [MDN Reference](https://developer.mozilla.org/docs/Web/CSS/:last-child)
 */
export const notLast = createVariant("&:not(:last-child)");

/**
 * Same as :nth-child(odd), Represents the odd elements of its parent.
 *
 * [MDN Reference](https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-child)
 */
export const odd = createVariant("&:nth-child(odd)");

/**
 * Same as :nth-child(even), Represents the even elements of its parent.
 *
 * [MDN Reference](https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-child)
 */
export const even = createVariant("&:nth-child(even)");

export const {
  visited,
  checked,
  disabled,
  focusWithin,
  hover,
  focus,
  focusVisible,
  active,
  link,
  target,
  notChecked,
  enabled,
  indeterminate,
  invalid,
  valid,
  optional,
  required,
  placeholderShown,
  readOnly,
  readWrite,
  notDisabled,
  firstOfType,
  notFirstOfType,
  lastOfType,
  notLastOfType,
  onlyChild,
  notOnlyChild,
  onlyOfType,
  notOnlyOfType,
  root,
  empty,
} = useProxy<PseudoClassVariants, Function>(prop => {
  prop = camelToDash(prop);
  if (prop.startsWith("not-")) prop = `not(:${prop.slice(4)})`;
  return createVariant("&:" + prop);
});
