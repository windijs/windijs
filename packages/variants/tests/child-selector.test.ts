import { all, children, sibling, siblings, svg } from "../src/childSelector";
import { bundle, prop } from "@windi/helpers";
import { colorHandler, configHandler, createUtility } from "@windi/core";

import { backgroundClipConfig } from "@windi/config";
import { windiColors } from "@windi/colors";

const bg = createUtility("bg")
  .use(colorHandler(windiColors, "backgroundColor", "--w-bg-opacity"))
  .case("clip", configHandler(backgroundClipConfig, ["backgroundClip", prop`-webkit-background-clip`]))
  .init();

const utilities = [bg.blue[500], bg.clip.content];

test("svg", () => {
  expect(bundle(svg(...utilities))).toMatchSnapshot();
});

test("all", () => {
  expect(bundle(all(...utilities))).toMatchSnapshot();
});

test("children", () => {
  expect(bundle(children(...utilities))).toMatchSnapshot();
});

test("siblings", () => {
  expect(bundle(siblings(...utilities))).toMatchSnapshot();
});

test("sibling", () => {
  expect(bundle(sibling(...utilities))).toMatchSnapshot();
});
