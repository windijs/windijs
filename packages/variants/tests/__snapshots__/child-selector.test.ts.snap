// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all 1`] = `
Map {
  "& *" => Object {
    "--w-bg-opacity": "1",
    "-webkit-background-clip": "content-box",
    "backgroundClip": "content-box",
    "backgroundColor": "rgba(59, 130, 246, var(--w-bg-opacity))",
  },
}
`;

exports[`children 1`] = `
Map {
  "& > *" => Object {
    "--w-bg-opacity": "1",
    "-webkit-background-clip": "content-box",
    "backgroundClip": "content-box",
    "backgroundColor": "rgba(59, 130, 246, var(--w-bg-opacity))",
  },
}
`;

exports[`sibling 1`] = `
Map {
  "& + *" => Object {
    "--w-bg-opacity": "1",
    "-webkit-background-clip": "content-box",
    "backgroundClip": "content-box",
    "backgroundColor": "rgba(59, 130, 246, var(--w-bg-opacity))",
  },
}
`;

exports[`siblings 1`] = `
Map {
  "& ~ *" => Object {
    "--w-bg-opacity": "1",
    "-webkit-background-clip": "content-box",
    "backgroundClip": "content-box",
    "backgroundColor": "rgba(59, 130, 246, var(--w-bg-opacity))",
  },
}
`;

exports[`svg 1`] = `
Map {
  "& svg" => Object {
    "--w-bg-opacity": "1",
    "-webkit-background-clip": "content-box",
    "backgroundClip": "content-box",
    "backgroundColor": "rgba(59, 130, 246, var(--w-bg-opacity))",
  },
}
`;
